src/tactic/equiv_rw.lean

```diff
@@ -200,12 +205,19 @@ do
   -- attempting to replace all occurrences of `e x`,
   -- calling it for now `j : β`, with `k : x = e.symm j`.
   generalize ex (by apply_opt_param) transparency.none,
-  -- Reintroduce `x` (now of type `b`).
+  -- Reintroduce `x` (now of type `b`), and the hypothesis `h`.
   intro x,
-  k ← mk_fresh_name,
-  -- Finally, we subst along `k`, hopefully removing all the occurrences of the original `x`,
-  intro k >>= (λ k, subst k <|> unfreeze_local_instances >> subst k),
-  `[try { simp only [equiv.symm_symm, equiv.apply_symm_apply, equiv.symm_apply_apply] }],
+  h ← mk_fresh_name >>= intro,
+  -- We may need to unfreeze `x` before we can `subst` or `clear` it.
+  unfreeze x',
+  -- Finally, if we're working on properties, substitute along `h`, then do some cleanup,
+  -- and if we're working on data, just throw out the old `x`.
+  b ← target >>= is_prop,
+  if b then (do
+    subst h,
+    `[try { simp only [] with equiv_rw_simp }])
+  else
+    clear' tt (native.rb_map.set_of_list [x']),
```

Comment:
What if `clear'` fails?

Reply:
I added an error message and failed. I think as this is currently outside the tested scope of `equiv_rw`, I'd prefer to fail hard rather than "best attempt" continue.